toolchain/check/node_id_traversal.cpp

```diff
@@ -0,0 +1,172 @@
+    // chunk of the parse tree.
+    if (chunks_.back().it == chunks_.back().end) {
+      auto old_chunk = chunks_.pop_back_val();
+
+      // If we're out of chunks, then we're done entirely.
+      if (chunks_.empty()) {
+        worklist_.VerifyEmpty();
+        return std::nullopt;
+      }
+
+      next_deferred_definition_.SkipTo(old_chunk.next_definition);
+      continue;
+    }
+
+    auto node_id = *chunks_.back().it;
+
+    // If we've reached the start of a deferred definition, skip to the end of
+    // it, and track that we need to check it later.
+    if (node_id == next_deferred_definition_.start_id()) {
+      const auto& definition_info =
+          context_.parse_tree().deferred_definitions().Get(
+              next_deferred_definition_.index());
+      worklist_.SuspendFunctionAndPush(context_,
+                                       next_deferred_definition_.index(),
+                                       definition_info.start_id);
+
+      // Continue type-checking the parse tree after the end of the definition.
+      chunks_.back().it =
+          Parse::Tree::PostorderIterator(definition_info.definition_id) + 1;
+      next_deferred_definition_.SkipTo(definition_info.next_definition_index);
+      continue;
+    }
+
+    ++chunks_.back().it;
+    return node_id;
+  }
+}
+
+// Determines whether this node kind is the start of a deferred definition
+// scope.
+static auto IsStartOfDeferredDefinitionScope(Parse::NodeKind kind) -> bool {
+  switch (kind) {
+    case Parse::NodeKind::ClassDefinitionStart:
+    case Parse::NodeKind::ImplDefinitionStart:
+    case Parse::NodeKind::InterfaceDefinitionStart:
+    case Parse::NodeKind::NamedConstraintDefinitionStart:
+      // TODO: Mixins.
+      return true;
+    default:
+      return false;
+  }
+}
+
+// Determines whether this node kind is the end of a deferred definition scope.
+static auto IsEndOfDeferredDefinitionScope(Parse::NodeKind kind) -> bool {
+  switch (kind) {
+    case Parse::NodeKind::ClassDefinition:
+    case Parse::NodeKind::ImplDefinition:
+    case Parse::NodeKind::InterfaceDefinition:
+    case Parse::NodeKind::NamedConstraintDefinition:
+      // TODO: Mixins.
+      return true;
+    default:
+      return false;
+  }
+}
+
+auto NodeIdTraversal::Handle(Parse::NodeKind parse_kind) -> void {
+  // When we reach the start of a deferred definition scope, add a task to the
+  // worklist to check future skipped definitions in the new context.
+  if (IsStartOfDeferredDefinitionScope(parse_kind)) {
+    worklist_.PushEnterDeferredDefinitionScope(context_);
+  }
+
+  // When we reach the end of a deferred definition scope, add a task to the
+  // worklist to leave the scope. If this is not a nested scope, start
+  // checking the deferred definitions now.
+  if (IsEndOfDeferredDefinitionScope(parse_kind)) {
+    chunks_.back().checking_deferred_definitions =
+        worklist_.SuspendFinishedScopeAndPush(context_);
+  }
```

Comment:
These calls make me sad for the same reason as the one in the worklist. Can you add a TODO for these also?

Reply:
Done